README.md
<!--
 * @Author: qzw w_qinzhiwei@xiwang.com
 * @Date: 2023-07-12 14:44:50
 * @LastEditors: qzw w_qinzhiwei@xiwang.com
 * @LastEditTime: 2023-07-13 11:13:25
 * @FilePath: /changelog-demo-1/README.md
 * @Description: 这是默认设置,请设置`customMade`, 打开koroFileHeader查看配置 进行设置: https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE
-->
# changelog-demo-1
commitlint changelog

1. 校验commitlint规则
2. 校验规则 subject-case camel-case
2. 校验规则 subject-case pascal-case PascalCase
2. 校验规则 subject-case lower-case default
2. 校验规则 subject-case
3. 测试changelog1
4. feat
5. 文档
6. style
7. refactor
8. perf
9. 构建，版本上线
10. test
11. feat
12. build
13. fix
14. feat
15. test
16. feat


CHANGELOG.md
# [1.6.0](https://github.com/qinzhiwei1993/changelog-demo-1/compare/v1.5.0...v1.6.0) (2023-07-13)


### ✅ Tests | 测试

* test ([ec6a255](https://github.com/qinzhiwei1993/changelog-demo-1/commit/ec6a2553215e2ab5a2529d623082560635fd3711))


### 🎫 Chores | 其他更新

* version ([731627d](https://github.com/qinzhiwei1993/changelog-demo-1/commit/731627db9047233797bcd1e92bfb7be68796c417))



# [1.5.0](https://github.com/qinzhiwei1993/changelog-demo-1/compare/v1.4.0...v1.5.0) (2023-07-12)


### ✨ Features | 新功能

* feat新功能 ([8f4342f](https://github.com/qinzhiwei1993/changelog-demo-1/commit/8f4342ffa248acb8c071033c4476408cb73c38a9))
* version升级 ([a8102a3](https://github.com/qinzhiwei1993/changelog-demo-1/commit/a8102a37b699dc366e3e3366c1902febc81f32f4))


### 📝 Documentation | 文档

* changelog ([fa0feaa](https://github.com/qinzhiwei1993/changelog-demo-1/commit/fa0feaa4159c1dae57deecd138d3b7da17755f8e))



# [1.4.0](https://github.com/qinzhiwei1993/changelog-demo-1/compare/v1.3.0...v1.4.0) (2023-07-12)


### ✨ Features | 新功能

* feat新功能 ([a21a180](https://github.com/qinzhiwei1993/changelog-demo-1/commit/a21a1804d5a1680c368f40276d0dc4d69f3ddd6b))
* test ([db1f590](https://github.com/qinzhiwei1993/changelog-demo-1/commit/db1f5909196612f04138e08ef3c4d211f552bd0d))
* version升级 ([c166a56](https://github.com/qinzhiwei1993/changelog-demo-1/commit/c166a564f8e4b4c5258a236e0ad4260b9b53d872))


### 🐛 Bug Fixes | Bug 修复

* 修复bug ([4c515c0](https://github.com/qinzhiwei1993/changelog-demo-1/commit/4c515c085c33d9e805ff21027db5770b6902e2c3)), closes [#123](https://github.com/qinzhiwei1993/changelog-demo-1/issues/123)


### 👷‍ Build System | 构建

* build 构建优化 ([5e1bf8c](https://github.com/qinzhiwei1993/changelog-demo-1/commit/5e1bf8c0503df8495a19bfcd05c48d8cf61cd35d))



# [1.3.0](https://github.com/qinzhiwei1993/changelog-demo-1/compare/v1.2.0...v1.3.0) (2023-07-12)


### ♻ Code Refactoring | 代码重构

* 项目重构 ([5c27fa4](https://github.com/qinzhiwei1993/changelog-demo-1/commit/5c27fa4450a223d641109bf776e3aa86d93ccb85))


### ⚡ Performance Improvements | 性能优化

* 性能优化 ([3011f60](https://github.com/qinzhiwei1993/changelog-demo-1/commit/3011f602dd575b75a706aa4205dfa5b94aac42fb))


### ✨ Features | 新功能

* 新增功能 ([28460ee](https://github.com/qinzhiwei1993/changelog-demo-1/commit/28460eec67816a897363f3c00f62d19feca9f611))


### 🎫 Chores | 其他更新

* 版本上线 ([74c58d4](https://github.com/qinzhiwei1993/changelog-demo-1/commit/74c58d4e354de91db2fe7a2b54f85a71dfad3e29))


### 🐛 Bug Fixes | Bug 修复

* 修复某个bug ([899f9b0](https://github.com/qinzhiwei1993/changelog-demo-1/commit/899f9b0c4d7b5c4bcdf5608c963d8ed2a2e65109))


### 💄 Styles | 风格

* 格式变更 ([b1f9dbc](https://github.com/qinzhiwei1993/changelog-demo-1/commit/b1f9dbc21e3e4b5582ecfe43838d35c9ec1cd088))


### 📝 Documentation | 文档

* changelog ([8dfb812](https://github.com/qinzhiwei1993/changelog-demo-1/commit/8dfb812d401e61af6309cdd9872198a72926ecc4))
* 文档变更 ([544a119](https://github.com/qinzhiwei1993/changelog-demo-1/commit/544a11967417504c921350152090733afbe9a9d0))


### BREAKING CHANGES

* ui库替换



# [1.2.0](https://github.com/qinzhiwei1993/changelog-demo-1/compare/v1.1.0...v1.2.0) (2023-07-12)


### ✨ Features | 新功能

* package.json版本升级 ([5d63a11](https://github.com/qinzhiwei1993/changelog-demo-1/commit/5d63a11d786b764dff83ff1eaa3745202a61399b))
* 移除拦截gitcommit ([66d6391](https://github.com/qinzhiwei1993/changelog-demo-1/commit/66d63913e691fffac4f244b47372e906cd933366))


### 📝 Documentation | 文档

* changelog ([e4e9841](https://github.com/qinzhiwei1993/changelog-demo-1/commit/e4e9841873ea907d538da6cc82bee0bf55a54d55))



# [1.1.0](https://github.com/qinzhiwei1993/changelog-demo-1/compare/v1.0.0...v1.1.0) (2023-07-12)


### ✨ Features | 新功能

* 测试chagenlog ([4776ddf](https://github.com/qinzhiwei1993/changelog-demo-1/commit/4776ddf6bcb2a114e06ae54465871fd14204a1f1))


### 📝 Documentation | 文档

* changelog ([c8aa455](https://github.com/qinzhiwei1993/changelog-demo-1/commit/c8aa455bcf786446ab7b2b4fa03d87d7377a60f9))



# [1.0.0](https://github.com/qinzhiwei1993/changelog-demo-1/compare/dd8a4c9b33c80d96728d0e0f46e2b5a77d48829a...v1.0.0) (2023-07-12)


### ✅ Tests | 测试

* 不能开头 Sentence case ([40bd7a6](https://github.com/qinzhiwei1993/changelog-demo-1/commit/40bd7a6891d7c8a4d187b736b9ea74317b53cb28))
* 校验规则 default ([1d8edeb](https://github.com/qinzhiwei1993/changelog-demo-1/commit/1d8edebafd7e8ce42784ea5074e20b996c2d650d))
* 校验规则 default ([15b6efa](https://github.com/qinzhiwei1993/changelog-demo-1/commit/15b6efa0a7193c68d0cfd393e914e8f790673b6c))
* 校验规则 default ([da95c10](https://github.com/qinzhiwei1993/changelog-demo-1/commit/da95c105d62c9dd5ae468214ff5aa2b1c5fc06a5))
* 校验规则cameCase ([94df213](https://github.com/qinzhiwei1993/changelog-demo-1/commit/94df213455e374d200532424a964f7a443f20ced))
* 校验规则PascalCase ([e0d9119](https://github.com/qinzhiwei1993/changelog-demo-1/commit/e0d9119bb82f390601f95b24101652a4ec0c8866))
* 校验规则UPPDERCASE ([fda271b](https://github.com/qinzhiwei1993/changelog-demo-1/commit/fda271be590994b985be260a0d1d5596e9cfafba))
* 格式校验 UPPERCASE ([b4b7929](https://github.com/qinzhiwei1993/changelog-demo-1/commit/b4b79290319f28425ea395eb098f02720d91a532))


### ✨ Features | 新功能

* 打tag v1.1.0 ([90628bb](https://github.com/qinzhiwei1993/changelog-demo-1/commit/90628bbb17c13ebb692d200745c1f82ad63eff74))
* 生成changelog ([12ea005](https://github.com/qinzhiwei1993/changelog-demo-1/commit/12ea0053b23dc7f1e339af1b5a7ddd0b50ce3ddc))


### 👷‍ Build System | 构建

* commitizen在提交的时候自动弹出 ([5e99c3e](https://github.com/qinzhiwei1993/changelog-demo-1/commit/5e99c3e4b881f84d593a3de4e8673ffa4224c55e))
* 初始化安装husky ([e332b1f](https://github.com/qinzhiwei1993/changelog-demo-1/commit/e332b1f5600d51244292cf97cead509851f62336))
* 安装commitizen 交互式提交commit ([5820d4a](https://github.com/qinzhiwei1993/changelog-demo-1/commit/5820d4a045943f4829ec9e5a04f6f8d803249b2c))
* 安装commitlint ([8216615](https://github.com/qinzhiwei1993/changelog-demo-1/commit/8216615cfb7a009fc1f0a8593d7ddb9362428b61))

changelog-options.js
/*
 * @Author: qzw w_qinzhiwei@xiwang.com
 * @Date: 2023-07-12 14:52:36
 * @LastEditors: qzw w_qinzhiwei@xiwang.com
 * @LastEditTime: 2023-07-12 14:52:45
 * @FilePath: /changelog-demo-1/changelog-options.js
 * @Description: 这是默认设置,请设置`customMade`, 打开koroFileHeader查看配置 进行设置: https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE
 */

/****
 * 
 * conventional-changelog自定义配置选项
 * 
 * 使用方式: conventional-changelog -c changelog-options.js -i CHANGELOG.md -s -r 0 
 */

const compareFunc = require('compare-func')

module.exports = {
    writerOpts: {
        transform: (commit, context) => {
            let discard = true
            const issues = []
            
            commit.notes.forEach(note => {
                note.title = 'BREAKING CHANGES'
                discard = false
            })
            if (commit.type === 'feat') {
                commit.type = '✨ Features | 新功能'
            } else if (commit.type === 'fix') {
                commit.type = '🐛 Bug Fixes | Bug 修复'
            } else if (commit.type === 'perf') {
                commit.type = '⚡ Performance Improvements | 性能优化'
            } else if (commit.type === 'revert' || commit.revert) {
                commit.type = '⏪ Reverts | 回退'
            } else if (false) { //  discard 后续不在输出到changelog中
                return
            } else if (commit.type === 'docs') {
                commit.type = '📝 Documentation | 文档'
            } else if (commit.type === 'style') {
                commit.type = '💄 Styles | 风格'
            } else if (commit.type === 'refactor') {
                commit.type = '♻ Code Refactoring | 代码重构'
            } else if (commit.type === 'test') {
                commit.type = '✅ Tests | 测试'
            } else if (commit.type === 'build') {
                commit.type = '👷‍ Build System | 构建'
            } else if (commit.type === 'ci') {
                commit.type = '🔧 Continuous Integration | CI 配置'
            } else if (commit.type === 'chore') {
                commit.type = '🎫 Chores | 其他更新'
            }

            if (commit.scope === '*') {
                commit.scope = ''
            }
            if (typeof commit.hash === 'string') {
                commit.shortHash = commit.hash.substring(0, 7)

            }
            if (typeof commit.subject === 'string') {
                let url = context.repository
                    ? `${context.host}/${context.owner}/${context.repository}`
                    : context.repoUrl
                if (url) {
                    url = `${url}/issues/`
                    // Issue URLs.
                    commit.subject = commit.subject.replace(/#([0-9]+)/g, (_, issue) => {
                        issues.push(issue)
                        return `[#${issue}](${url}${issue})`
                    })
                }
                if (context.host) {
                    // User URLs.
                    commit.subject = commit.subject.replace(/\B@([a-z0-9](?:-?[a-z0-9/]){0,38})/g, (_, username) => {
                        if (username.includes('/')) {
                            return `@${username}`
                        }

                        return `[@${username}](${context.host}/${username})`
                    })
                }
            }

            // remove references that already appear in the subject
            commit.references = commit.references.filter(reference => {
                if (issues.indexOf(reference.issue) === -1) {
                    return true
                }

                return false
            })
            return commit
        },
        groupBy: 'type',
        commitGroupsSort: 'title',
        commitsSort: ['scope', 'subject'],
        noteGroupsSort: 'title',
        notesSort: compareFunc
    }
}


commitlint.config.js
/*
 * @Author: qzw w_qinzhiwei@xiwang.com
 * @Date: 2023-07-12 11:17:40
 * @LastEditors: qzw w_qinzhiwei@xiwang.com
 * @LastEditTime: 2023-07-12 14:26:53
 * @FilePath: /changelog-demo-1/commitlint.config.js
 * @Description: 这是默认设置,请设置`customMade`, 打开koroFileHeader查看配置 进行设置: https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE
 */
module.exports = {
    extends: ['@commitlint/config-conventional'],
    rules: {
        'body-leading-blank': [1, 'always'],
        'body-max-line-length': [2, 'always', 100],
        'footer-leading-blank': [1, 'always'],
        'footer-max-line-length': [2, 'always', 100],
        'header-max-length': [2, 'always', 100],
        'subject-case': [
            2,
            'never',
            ['sentence-case', 'start-case', 'pascal-case', 'upper-case'],
        ],
        'subject-empty': [2, 'never'],
        'subject-full-stop': [2, 'never', '.'],
        'type-case': [2, 'always', 'lower-case'],
        'type-empty': [2, 'never'],
        'type-enum': [
            2,
            'always',
            [
                'build',
                'chore',
                'ci',
                'docs',
                'feat',
                'fix',
                'perf',
                'refactor',
                'revert',
                'style',
                'test',
            ],
        ],
    },
    prompt: {
		questions: {
			type: {
				description: "Select the type of change that you're committing",
				enum: {
					feat: {
						description: 'A new feature',
						title: 'Features',
						emoji: '✨',
					},
					fix: {
						description: 'A bug fix',
						title: 'Bug Fixes',
						emoji: '🐛',
					},
					docs: {
						description: 'Documentation only changes',
						title: 'Documentation',
						emoji: '📚',
					},
					style: {
						description:
							'Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)',
						title: 'Styles',
						emoji: '💎',
					},
					refactor: {
						description:
							'A code change that neither fixes a bug nor adds a feature',
						title: 'Code Refactoring',
						emoji: '📦',
					},
					perf: {
						description: 'A code change that improves performance',
						title: 'Performance Improvements',
						emoji: '🚀',
					},
					test: {
						description: 'Adding missing tests or correcting existing tests',
						title: 'Tests',
						emoji: '🚨',
					},
					build: {
						description:
							'Changes that affect the build system or external dependencies (example scopes: gulp, broccoli, npm)',
						title: 'Builds',
						emoji: '🛠',
					},
					ci: {
						description:
							'Changes to our CI configuration files and scripts (example scopes: Travis, Circle, BrowserStack, SauceLabs)',
						title: 'Continuous Integrations',
						emoji: '⚙️',
					},
					chore: {
						description: "Other changes that don't modify src or test files",
						title: 'Chores',
						emoji: '♻️',
					},
					revert: {
						description: 'Reverts a previous commit',
						title: 'Reverts',
						emoji: '🗑',
					},
				},
			},
			scope: {
				description:
					'What is the scope of this change (e.g. component or file name)',
			},
			subject: {
				description:
					'Write a short, imperative tense description of the change',
			},
			body: {
				description: 'Provide a longer description of the change',
			},
			isBreaking: {
				description: 'Are there any breaking changes?',
			},
			breakingBody: {
				description:
					'A BREAKING CHANGE commit requires a body. Please enter a longer description of the commit itself',
			},
			breaking: {
				description: 'Describe the breaking changes',
			},
			isIssueAffected: {
				description: 'Does this change affect any open issues?',
			},
			issuesBody: {
				description:
					'If issues are closed, the commit requires a body. Please enter a longer description of the commit itself',
			},
			issues: {
				description: 'Add issue references (e.g. "fix #123", "re #123".)',
			},
		},
	}
}
